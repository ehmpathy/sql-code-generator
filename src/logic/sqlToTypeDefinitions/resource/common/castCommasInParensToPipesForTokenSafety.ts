export const castCommasInParensToPipesForTokenSafety = ({ sql }: { sql: string }) => {
  // track state of whether we are in parenthesizes or not
  let areInsideParensDepth = 0; // init as 0, since not inside parens
  const areInsideParens = () => areInsideParensDepth > 0; // whenever areInsideParensDepth > 0, we are inside parens

  // init object to build up the "casted" string
  let castedString = '';

  // loop through each char and update state / replace char as needed
  [...sql].forEach((char) => {
    // 1. manage "areInsideParens" state
    if (char === '(') {
      areInsideParensDepth += 1; // one parens deeper!
    }
    if (char === ')') {
      if (!areInsideParens) throw new Error('exiting parenthesizes without having opened them; not supported'); // fail fast for unsupported case
      areInsideParensDepth -= 1; // mark that we've dropped off one parens now
    }

    // 2. cast the char, considering whether we're in parens or not
    const castedChar = areInsideParens() && char === ',' ? '|' : char; // cast "," if we are inside parens

    // 3. add casted char to string
    castedString += castedChar;
  });

  // return the casted string
  return castedString;
};
