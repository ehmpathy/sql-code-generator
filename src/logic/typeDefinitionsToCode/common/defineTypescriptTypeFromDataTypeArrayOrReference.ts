import {
  DataType,
  TypeDefinitionReference,
  TypeDefinition,
  TypeDefinitionOfQueryTableReference,
} from '../../../domain';
import { defineTypescriptTypeFromDataTypeArray } from './defineTypescriptTypeFromDataTypeArray';
import { defineTypescriptTypeFromReference } from './defineTypescriptTypeFromReference/defineTypescriptTypeFromReference';

export const defineTypescriptTypeFromDataTypeArrayOrReference = ({
  type,
  plural,
  queryTableReferences,
  typeDefinitions,
}: {
  type: DataType[] | TypeDefinitionReference;
  plural: boolean;
  typeDefinitions: TypeDefinition[];
  queryTableReferences: TypeDefinitionOfQueryTableReference[];
}) => {
  // if its a type reference, then use that handler
  if (type instanceof TypeDefinitionReference) {
    const defined = defineTypescriptTypeFromReference({
      reference: type,
      queryTableReferences,
      typeDefinitions,
    });
    if (plural) return `${defined}[]`;
    return defined;
  }

  // else, it must be data array. use that handler
  return defineTypescriptTypeFromDataTypeArray({ type });
};
