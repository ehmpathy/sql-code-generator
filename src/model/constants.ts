import {
  TypeDefinitionOfQuery,
  TypeDefinitionOfResourceFunction,
  TypeDefinitionOfResourceTable,
  TypeDefinitionOfResourceView,
} from './valueObjects';

export enum DatabaseLanguage {
  MYSQL = 'mysql',
  POSTGRES = 'postgres',
}
export enum DefinitionType {
  RESOURCE = 'resource',
  QUERY = 'query',
}
export enum DataType {
  STRING = 'string',
  STRING_ARRAY = 'string[]', // postgres supports arrays
  NUMBER = 'number',
  NUMBER_ARRAY = 'number[]', // postgres supports arrays
  DATE = 'Date',
  BUFFER = 'Buffer', // e.g., for binary
  NULL = 'null',
  UNDEFINED = 'undefined',
}
export enum ResourceType {
  TABLE = 'TABLE',
  FUNCTION = 'FUNCTION',
  PROCEDURE = 'PROCEDURE',
  VIEW = 'VIEW',
}
export enum QuerySection {
  SELECT_EXPRESSIONS = 'SELECT_EXPRESSIONS',
  TABLE_REFERENCES = 'TABLE_REFERENCES',
  WHERE_CONDITIONS = 'WHERE_CONDITIONS',
}
export type TypeDefinition =
  | TypeDefinitionOfQuery
  | TypeDefinitionOfResourceTable
  | TypeDefinitionOfResourceFunction
  | TypeDefinitionOfResourceView;
export interface GeneratedOutputPaths {
  types: string;
  queryFunctions: string;
}
