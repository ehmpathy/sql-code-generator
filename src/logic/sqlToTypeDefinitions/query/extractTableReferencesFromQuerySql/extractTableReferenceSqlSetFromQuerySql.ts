import { TABLE_REFERENCE_TYPE } from './constants';

/*
  TODO: handle subqueries: https://github.com/uladkasach/sql-code-generator/issues/2
*/
export const extractTableReferenceSqlSetFromQuerySql = ({ sql }: { sql: string }) => {
  // 1. grab the content of everything between FROM and WHERE
  const everythingAfterFromInclusive = (() => {
    const partsSplitOnFrom = sql.split(/[\n\r\s]+from[\n\r\s]+/gi);
    if (partsSplitOnFrom.length === 1) throw new Error('no "from" keyword found'); // fail fast; allow this being caught above
    if (partsSplitOnFrom.length > 2) throw new Error('more than one "from" keyword found; not yet supported'); // TODO: https://github.com/uladkasach/sql-code-generator/issues/2
    return `FROM ${partsSplitOnFrom[1]}`; // inclusive, since we include "from"
  })();
  const everythingBetweenFromAndWhere = (() => {
    const partsSplitOnFromAfterSelect = everythingAfterFromInclusive.split(/(?:WHERE|where)/g);
    if (partsSplitOnFromAfterSelect.length > 2) {
      throw new Error('more than one "where" keyword found; not yet supported'); // TODO: https://github.com/uladkasach/sql-code-generator/issues/2
    }
    return partsSplitOnFromAfterSelect[0];
  })();

  // 3. split token on each table reference type "FROM", "JOIN", "INNER JOIN", and "LEFT JOIN"
  const tableReferenceDefinitionSplitRegex = new RegExp(
    `(?:[\\n\\r\\s]+)(${TABLE_REFERENCE_TYPE.join('|')}|${TABLE_REFERENCE_TYPE.map((str) => str.toLowerCase()).join(
      '|',
    )})(?:[\\n\\r\\s]+)`,
    'g',
  );
  const partsSplitOnJoinType = ` ${everythingBetweenFromAndWhere} `.split(tableReferenceDefinitionSplitRegex);
  const tableReferenceSqlSet = [] as string[];
  let partialTableReferenceSql: null | string = null;
  partsSplitOnJoinType.slice(1).forEach((part, index) => {
    if (index % 2 === 0) {
      // if even, then we should be starting the table reference sql
      if (!TABLE_REFERENCE_TYPE.includes(part.toUpperCase())) throw new Error('unexpected'); // fail fast, this should never occur
      if (partialTableReferenceSql !== null) throw new Error('unexpected'); // fail fast, this should never occur
      partialTableReferenceSql = part; // initialize the partial table reference
    } else {
      tableReferenceSqlSet.push(`${partialTableReferenceSql} ${part}`.trim().replace(';', ''));
      partialTableReferenceSql = null;
    }
  });

  // return the reference set
  return tableReferenceSqlSet;
};
