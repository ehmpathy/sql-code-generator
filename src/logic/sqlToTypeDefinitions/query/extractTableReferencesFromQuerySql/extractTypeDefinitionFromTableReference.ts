import { TypeDefinitionOfQueryTableReference } from '../../../../domain/objects/TypeDefinitionOfQueryTableReference';
import { TABLE_REFERENCE_TYPE } from './constants';

const TABLE_NAME_MATCHER_REGEX = `(?:${
  // upper case (e.g., "FROM', 'JOIN', 'LEFT JOIN')
  TABLE_REFERENCE_TYPE.join('|')
}|${
  // and lowercase (e.g., 'from', 'join', 'left join')
  TABLE_REFERENCE_TYPE.map((str) => str.toLowerCase()).join('|')
})\\s(?:\\w+\\.)?(\\w+)?(?:\\s|$)`;

const FUNCTION_NAME_MATCHER_REGEX = `(?:${
  // upper case (e.g., "FROM', 'JOIN', 'LEFT JOIN')
  TABLE_REFERENCE_TYPE.join('|')
}|${
  // and lowercase (e.g., 'from', 'join', 'left join')
  TABLE_REFERENCE_TYPE.map((str) => str.toLowerCase()).join('|')
})\\s(?:\\w+\\.)?(\\w+)?(?:\\()`;

const SPECIFIED_ALIAS_MATCHER_REGEX = `(?:${
  // upper case (e.g., "FROM', 'JOIN', 'LEFT JOIN')
  TABLE_REFERENCE_TYPE.join('|')
}|${
  // and lowercase (e.g., 'from', 'join', 'left join')
  TABLE_REFERENCE_TYPE.map((str) => str.toLowerCase()).join('|')
})\\s(?:\\w+\\.)?(?:\\w+)(?:\\([\\(\\|\\,\\s\\:\\_\\w\\)]*\\))?(?:\\s(?:as|AS))?\\s((?!ON|on)\\w+)?`;

export const extractTypeDefinitionFromTableReference = ({
  sql,
}: {
  sql: string;
}) => {
  // see if it matches a persisted table reference
  // eslint-disable-next-line @typescript-eslint/naming-convention, @typescript-eslint/no-unused-vars
  const [_, tableName] = new RegExp(TABLE_NAME_MATCHER_REGEX).exec(sql) ?? []; // tslint:disable-line no-unused

  // see if it matches a function output reference
  // eslint-disable-next-line @typescript-eslint/naming-convention, @typescript-eslint/no-unused-vars
  const [__, functionName] = tableName
    ? [, null]
    : new RegExp(FUNCTION_NAME_MATCHER_REGEX).exec(sql) ?? []; // tslint:disable-line no-unused

  // make sure it matches one of the above
  if (!tableName && !functionName)
    throw new Error(
      'could not identify the referenced table from table reference sql; unexpected',
    ); // fail fast

  // grab the explicit alias, if any
  // eslint-disable-next-line @typescript-eslint/naming-convention, @typescript-eslint/no-unused-vars
  const [___, specifiedAlias] =
    new RegExp(SPECIFIED_ALIAS_MATCHER_REGEX).exec(sql) ?? []; // tslint:disable-line no-unused

  // define the name, considering whether alias was given
  const alias: string = specifiedAlias ?? tableName ?? functionName!;

  // return the full definition
  return new TypeDefinitionOfQueryTableReference({
    alias,
    tableName: tableName ?? null,
    functionName: functionName ?? null,
  });
};
