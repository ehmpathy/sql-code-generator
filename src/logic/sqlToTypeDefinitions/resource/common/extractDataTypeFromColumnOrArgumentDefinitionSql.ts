import { DataType } from '../../../../model';

// https://dev.mysql.com/doc/refman/8.0/en/string-types.html
const mysqlStringTypes = [
  'CHAR',
  'VARCHAR',
  'TINYTEXT',
  'TEXT',
  'MEDIUMTEXT',
  'LONGTEXT',
  'ENUM', // note: we treat enums as strings; TODO: https://github.com/uladkasach/sql-code-generator/issues/1
];

// https://www.postgresql.org/docs/9.5/datatype-character.html
const postgresStringTypes = [
  'CHARACTER',
  'CHAR',
  'CHARACTER VARYING',
  'VARCHAR',
  'TEXT',
  'UUID', // this too: https://www.postgresql.org/docs/9.5/datatype-uuid.html
];

// combine the types so we can search over the union
const dbStringTypes = [...mysqlStringTypes, ...postgresStringTypes];

// https://dev.mysql.com/doc/refman/8.0/en/numeric-types.html
const mysqlNumberTypes = [
  'INTEGER',
  'INT',
  'SMALLINT',
  'TINYINT',
  'MEDIUMINT',
  'BIGINT',
  'DECIMAL',
  'NUMERIC',
  'FLOAT',
  'DOUBLE',
];

// https://www.postgresql.org/docs/9.5/datatype-numeric.html
const postgresNumberTypes = [
  'SMALLINT',
  'INT2',
  'INTEGER',
  'INT',
  'INT4',
  'BIGINT',
  'INT8',
  'DECIMAL',
  'NUMERIC',
  'REAL',
  'DOUBLE PRECISION',
  'FLOAT8',
  'SMALLSERIAL',
  'SERIAL2',
  'SERIAL',
  'SERIAL4',
  'BIGSERIAL',
  'SERIAL8',
];

// combine the types so we can search over the union
const dbNumberTypes = [...mysqlNumberTypes, ...postgresNumberTypes];

// https://dev.mysql.com/doc/refman/8.0/en/date-and-time-types.html
const mysqlDateTypes = ['DATE', 'TIME', 'DATETIME', 'TIMESTAMP', 'YEAR'];

// https://www.postgresql.org/docs/9.5/datatype-datetime.html
const postgresDateTypes = [
  'TIMESTAMP',
  'TIMESTAMPTZ',
  'TIMESTAMP WITH TIMEZONE',
  'DATE',
  'TIME',
  'TIMETZ',
  'TIME WITH TIMEZONE',
];

// combine the types so we can search over the union
const dbDateTypes = [...mysqlDateTypes, ...postgresDateTypes];

// https://dev.mysql.com/doc/refman/8.0/en/binary-varbinary.html
const mysqlBinaryTypes = ['BINARY', 'VARBINARY'];

// https://www.postgresql.org/docs/9.5/datatype-binary.html
const postgresBinaryTypes = ['BYTEA'];

// combine the types so we can search over the union
const dbBinaryTypes = [...mysqlBinaryTypes, ...postgresBinaryTypes];

export const extractDataTypeFromColumnOrArgumentDefinitionSql = ({ sql }: { sql: string }) => {
  // 1. cast to upper
  const sqlUpper = sql.toUpperCase();

  // 2. one by one try to see which datatype is being defined in this definition
  if (dbStringTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.STRING;
  if (dbNumberTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.NUMBER;
  if (dbDateTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.DATE;
  if (dbBinaryTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.BUFFER;
  throw new Error(`could not extract data type from '${sql}'`);
};
