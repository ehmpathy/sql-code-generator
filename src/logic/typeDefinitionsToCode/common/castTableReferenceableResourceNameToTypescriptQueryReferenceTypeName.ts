import { ResourceType } from '../../../model';
import { castResourceNameToTypescriptTypeName } from './castResourceNameToTypescriptTypeName';

export const castTableReferenceableResourceNameToTypescriptQueryReferenceTypeName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  // 0. throw error if not view or table, only view and table are possible table references
  if (![ResourceType.TABLE, ResourceType.VIEW].includes(resourceType)) {
    throw new Error('not a table referenceable query');
  }

  // 1. get the regular typescript type name
  const typescriptTypeName = castResourceNameToTypescriptTypeName({ name, resourceType });

  // 2. return the referenceable name
  if (resourceType === ResourceType.TABLE) return typescriptTypeName.replace(/^SqlTable/, 'SqlTableOrView');
  if (resourceType === ResourceType.VIEW) return typescriptTypeName.replace(/^SqlView/, 'SqlTableOrView');
  throw new Error('unexpected'); // this should never occur, but fail fast
};
