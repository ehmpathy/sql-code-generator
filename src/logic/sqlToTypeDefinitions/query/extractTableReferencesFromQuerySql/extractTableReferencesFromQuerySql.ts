import { extractTypeDefinitionFromTableReference } from './extractTypeDefinitionFromTableReference';
import { tryToExtractTableReferenceSqlSet } from './tryToExtractTableReferenceSqlSetFromQuerySql';
import { flattenSqlByReferencingAndTokenizingSubqueries } from '../common/flattenSqlByReferencingAndTokenizingSubqueries';

export const extractTableReferencesFromQuerySql = ({ sql }: { sql: string }) => {
  // flatten the sql and tokenize each subquery - otherwise subquery syntax throws off subsequent processing
  const { flattenedSql: rootQuerySql, references: subqueries } = flattenSqlByReferencingAndTokenizingSubqueries({
    sql,
  });

  // on both the root query sql and each subquery sql (if subqueries are present) extract table references
  const flattenedSqlsToConsider = [rootQuerySql, ...subqueries.map((subquery) => subquery.sql)];
  const tableReferences = flattenedSqlsToConsider
    .map((sql) => {
      // get the table reference sqls from the main query sql
      const tableReferenceSqlSet = tryToExtractTableReferenceSqlSet({ sql }); // "try to", since its okay if there are no table references; (e.g., `select upsert_something(...)`)

      // get the type definitions of table references from each table reference sql
      return tableReferenceSqlSet.map((sql) => extractTypeDefinitionFromTableReference({ sql }));
    })
    .flat();

  // return the table references
  return tableReferences;
};
