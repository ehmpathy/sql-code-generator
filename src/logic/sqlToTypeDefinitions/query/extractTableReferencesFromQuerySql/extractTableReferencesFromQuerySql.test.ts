import { extractSqlFromFile } from '../../../common/extractSqlFromFile';
import { extractTableReferencesFromQuerySql } from './extractTableReferencesFromQuerySql';
import { TypeDefinitionOfQueryTableReference } from '../../../../model';

describe('extractTableReferencesFromQuerySql', () => {
  it('should be able to determine types accurately for this example', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/find_image_by_id.sql`,
    });
    const defs = extractTableReferencesFromQuerySql({ sql });
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for this other example', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/select_suggestion.sql`,
    });
    const defs = extractTableReferencesFromQuerySql({ sql });
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for an example joining to multiple tables', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/find_users_by_last_name.sql`,
    });
    const defs = extractTableReferencesFromQuerySql({ sql });
    expect(defs.length).toEqual(3);
    expect(defs).toContainEqual(
      new TypeDefinitionOfQueryTableReference({ alias: 'up', tableName: 'view_user_profile_current' }),
    );
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for an example with subqueries referencing their own tables', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/find_providers_with_work_count.sql`,
    });
    const defs = extractTableReferencesFromQuerySql({ sql });
    expect(defs.length).toEqual(2);
    expect(defs).toContainEqual(new TypeDefinitionOfQueryTableReference({ alias: 'w', tableName: 'work' })); // has the subqueries definition
    expect(defs).toMatchSnapshot();
  });
});
