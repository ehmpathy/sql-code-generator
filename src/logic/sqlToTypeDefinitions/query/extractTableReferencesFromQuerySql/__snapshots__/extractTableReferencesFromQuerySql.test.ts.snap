// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractTableReferencesFromQuerySql should be able to determine types accurately for an example joining to multiple tables 1`] = `
Array [
  TypeDefinitionOfQueryTableReference {
    "alias": "u",
    "tableName": "user",
  },
  TypeDefinitionOfQueryTableReference {
    "alias": "p",
    "tableName": "phone",
  },
  TypeDefinitionOfQueryTableReference {
    "alias": "up",
    "tableName": "view_user_profile_current",
  },
]
`;

exports[`extractTableReferencesFromQuerySql should be able to determine types accurately for an example with subqueries referencing their own tables 1`] = `
Array [
  TypeDefinitionOfQueryTableReference {
    "alias": "p",
    "tableName": "provider",
  },
  TypeDefinitionOfQueryTableReference {
    "alias": "w",
    "tableName": "work",
  },
]
`;

exports[`extractTableReferencesFromQuerySql should be able to determine types accurately for this example 1`] = `
Array [
  TypeDefinitionOfQueryTableReference {
    "alias": "i",
    "tableName": "image",
  },
]
`;

exports[`extractTableReferencesFromQuerySql should be able to determine types accurately for this other example 1`] = `
Array [
  TypeDefinitionOfQueryTableReference {
    "alias": "s",
    "tableName": "suggestion",
  },
  TypeDefinitionOfQueryTableReference {
    "alias": "cvp",
    "tableName": "suggestion_cvp",
  },
  TypeDefinitionOfQueryTableReference {
    "alias": "v",
    "tableName": "suggestion_version",
  },
]
`;
